rust/kernel/mm.rs

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Memory management.
+//!
+//! C header: [`include/linux/mm.h`](../../../../include/linux/mm.h)
+
+use crate::{bindings, pages, to_result, Result};
+
+/// Virtual memory.
+pub mod virt {
+    use super::*;
+
+    /// A wrapper for the kernel's `struct vm_area_struct`.
+    ///
+    /// It represents an area of virtual memory.
+    ///
+    /// # Invariants
+    ///
+    /// `vma` is always non-null and valid.
+    pub struct Area {
+        vma: *mut bindings::vm_area_struct,
+    }
+
+    impl Area {
+        /// Creates a new instance of a virtual memory area.
+        ///
+        /// # Safety
+        ///
+        /// Callers must ensure that `vma` is non-null and valid for the duration of the new area's
+        /// lifetime.
+        pub(crate) unsafe fn from_ptr(vma: *mut bindings::vm_area_struct) -> Self {
+            // INVARIANTS: The safety requirements guarantee the invariants.
+            Self { vma }
+        }
```

Comment:
Maybe `Area` could get a lifetime parameter and then `from_ptr` could take an `dyn for<'a> FnOnce(Area<'a>) -> T`? This would ensure that the `Area` is dropped before `from_ptr` returns. I don't know if this would work for all uses of `Area` though.

Reply:
I think `Area` is one of the most complicated objects we'll have because its underlying object's (`struct vm_area_struct`) lifetime is not managed statically or with refcounts. Instead, we need to hold the `mm` lock in at least read mode while using the vma (to ensure that it doesn't go away), and when it does go away, a callback is called (`vm_area_struct::vm_ops::close`).

So we'll need to hold on to the owning `mm` (which is refcounted), and only grant access to the underlying `vma` when we know `mm` is locked (either because we lock it ourselves, or we know implicitly it's locked, or perhaps explicitly, e.g., by having callers provide a `Guard`), and that `close` was never called (it is called with the lock held, so if we're holding the lock, we know it won't be called).

If we don't hold on to `vma` outside the scope of the `mmap` callback, we don't care about the above. But in Binder, for example, we do care about it because we want to unmap and recycle pages when they're not in use and the system is under memory pressure. (Then we want to remap them if the system needs the area to perform IPCs.)

For another driver I'm looking at, we have to remember all `vma`s because we need to unmap them when a device is unbound from the driver.

Anyway, I don't think adding a lifetime to `Area` is helpful now because we'll likely need to remove it later when we cover the scenarios above.